### PR TITLE
Viki expanding vfind

### PR DESCRIPTION
**Pull request by lvtiendev**
Expanding vfind. return a json string of metadata (currently only blocked state) for each result returned from the query => double the number of lines returned
The additional param is a string which follows exclusive list param and before detail_field param in the command.
To fetch blocked item, use string "withblocked"
otherwise, use "noblocked"

### CODE REVIEW

File: src/vfind.c [hunk 1]
```diff
@@ -50,13 +57,14 @@ int *vfindGetKeys(struct redisCommand *cmd, robj **argv, int argc, int *numkeys,
 
 // Prepares all the sets(data, cap, anticap, filters) and calls vfindByFilters or vfindByZWithFilters
 // depending on the ratio of the input set and the smallest filter set
+// Syntax: vfind zset cap anticap offset count upto direction incl excl detail_field filter_count [filter keys]
```
Comment (karlseguin) [inline src/vfind.c]: missing the new include_blocked parameter in the syntax

**Comment by karlseguin** [comment]
This commit is hardcore.  Down the rabbit hole we go.

2 thoughts.

1 - I know it's more efficient, and I know we do it in other places, but the duplication of code between the core blocks based on the state of include_blocked is pretty extreme. I'm fine with it as-is, but talk it over with the team one more time. An if check on an integer doesn't seem like that high a price to pay for me.

2 - I'm uncomfortable with modifying redisObject. We've done a good job staying away from the core redis codebase, which has made it easy for us to manage our own fork. Why not create a vikiObject and encapsulate a redisObject within it? Extra allocation, I know, but in this case I think it's definitely worth the price.

**Comment by karlseguin** [comment]
Do we think it's reasonable to exposed blocked results out of vsort (used for auto complete) ?

**Comment by lvtiendev** [comment]
Hi Karl, thanks for the feedback.
1. ok I will try to rewrite the code and do some benchmark to compare the performance.
2. Actually I have tried with several implementations that minimize the change to redisObject. However, it seems like redis only allows us to persist redisObject and it treats all other custom objects as memory leak if you store them in the memory. I can look for where they check memory leak and see if we can modify/overwrite it. But then it still leads to problems with future redis update

**Comment by cviedmai** [comment]
I do agree that modifying the RedisObject is the most hardcore and undesirable approach.. but seems the best approach. 

On top of what Tony already said,  another problem with building our own VikiObject is that the existing SkipList implementation only works out with RedisObjects. Thus, we would have to create our own implementation of SkipList in order to insert items there in order. Not like it's a lot of work (maybe a hundred lines of code?) but it's yet another thing to maintain.

Also, Tony tried building a metadata struct to store this info (which would be very nice to extend functionality as well) and have it associated with each of our RedisObjects via a HashMap (the keys being the RedisObject itself, and the values the metadata), but this added an extra 1.2ms (50% slower) for a vfind of 25 results. More results it was even slower. The fastest approach is modifying the redisObject which only have a penalty of 300 us for the basic vfind.

**Comment by karlseguin** [comment]
It looks good to me. We'll need to coordinate a deploy of both Hyperion and Legacy.

**Comment by cviedmai** [comment]
Let's do it when you're back next week.

**Pull request merged**